### PR TITLE
Normal docs and tweaks

### LINKED ISSUE #325: Start on rng algorithms doc
Writing the ziggurat algorithm was unpleasant, add a vignette that explains how it is derived and keep track of any optimisations we add as they mostly reduce readability

### PR DESCRIPTION
**Pull request by richfitz**
Documentation-driven development: 

* New polar algorithm for normally distributed random numbers; faster than Box-Muller but slower than Ziggurat
* Tweaked algorithm for Ziggurat so now ~10% faster by avoiding drawing a second random number when working with 64 bit integers
* Added a new vignette describing the internals of the normal sampling algorithms

Fixes #325

### CODE REVIEW

File: inst/include/dust/random/normal_ziggurat.hpp [hunk 1]
```diff
@@ -28,6 +28,20 @@ real_type normal_ziggurat_tail(rng_state_type& rng_state, real_type x1,
   } while (true);
   return ret;
 }
+
+template <typename rng_state_type>
+int ziggurat_layer(rng_state_type& rng_state,
```
Comment (johnlees) [inline inst/include/dust/random/normal_ziggurat.hpp:33]: I think `ziggurat_layer_draw` might be a little clearer as a name. Also a quick note to say what this is doing, as bitshifting is always a bit opaque

File: vignettes/rng_normal.Rmd [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+n <- 1e6
+bench::mark(
+  r = rnorm(n),
+  box_muller = rng$random_normal(n, algorithm = "box_muller"),
+  polar = rng$random_normal(n, algorithm = "polar"),
+  ziggurat = rng$random_normal(n, algorithm = "ziggurat"),
+  check = FALSE, time_unit = "ms")
+```
+
+(This is not their intended purpose of course - that is to be called from C++ in parallel.)  This document collects some notes on the algorithms that underly the code used.
```
Comment (johnlees) [inline vignettes/rng_normal.Rmd:39]:
```suggestion
(This is not their intended purpose of course - that is to be called from C++ in parallel.)  This document collects some notes on the algorithms that underlie the code used.
```

**Review by johnlees — APPROVED** [review]
Looks like you had fun writing this! Very good and detailed docs

File: inst/include/dust/random/normal_ziggurat.hpp (hunk 1 above)
Comment (richfitz) [reply inst/include/dust/random/normal_ziggurat.hpp]: Agreed on both counts, have added a bit of a wall of text now

File: vignettes/rng_normal.Rmd (hunk 2 above)
Comment (richfitz) [reply vignettes/rng_normal.Rmd]: Done, among quite a  few spelling mistakes

**Pull request merged**